File: app/static/js/neoscorpion_fuel_planning.js
(() => {
    "use strict";

    const planThreeTank = (required, remaining, wingMax, wingThreshold) => {
        const centerRemaining = Number(remaining.ctr || 0);
        let left;
        if (centerRemaining > 0) {
            left = required >= wingThreshold + centerRemaining
                ? wingMax
                : (required - centerRemaining) / 2;
        } else {
            left = required >= wingThreshold ? wingMax : required / 2;
        }
        const center = required <= wingThreshold
            ? centerRemaining
            : required - left - left;
        return {left, ctr: center, right: left};
    };

    const basePlan = (aircraftType, required, remaining = {}, actual = {}) => {
        if (!Number.isFinite(required)) return null;
        if (aircraftType === "B757") {
            return planThreeTank(required, remaining, 14.6, 29.2);
        }
        if (aircraftType === "B767ER") {
            return planThreeTank(required, remaining, 40.2, 80.4);
        }
        if (aircraftType === "A300") {
            const centerRemaining = Number(remaining.ctr || 0);
            const outboard = required >= 16.4 ? 8.2 : (required - 4.1) / 2;
            let inboard;
            if (centerRemaining > 0 && required < 78.6) {
                inboard = ((required - outboard - outboard) / 2)
                    - (centerRemaining / 2);
            } else {
                inboard = required >= 78.6
                    ? 31.1
                    : (required - outboard - outboard) / 2;
            }
            const center = required >= 109.7
                ? 31.1
                : required - outboard - inboard - inboard - outboard;
            const trim = required <= 109.7
                ? 0
                : required - outboard - inboard - center - inboard - outboard;
            return {
                l_out: outboard,
                l_in: inboard,
                ctr: center,
                r_in: inboard,
                r_out: outboard,
                tt: trim,
            };
        }
        if (aircraftType === "B747-400") {
            const leftOutboard = required >= 117.168 ? 29.292 : required / 4;
            const leftInboard = required >= 244.41
                ? 84.058
                : Number(actual.main_l_in || 0);
            const mainTotal = (leftOutboard + leftInboard) * 2;
            const reserve = required >= 163.51
                ? 8.857
                : (required - mainTotal) / 2;
            return {
                main_l_out: leftOutboard,
                main_l_in: leftInboard,
                main_r_in: leftInboard,
                main_r_out: leftOutboard,
                reserve_2_l: reserve,
                reserve_3_r: reserve,
                center_wing: required - mainTotal - reserve - reserve,
            };
        }
        if (aircraftType === "B747-8") {
            const leftOutboard = required >= 183.688
                ? 35.644
                : Number(actual.main_l_out || 0);
            const leftInboard = required >= 285.206
                ? 96.681
                : Number(actual.main_l_in || 0);
            const reserve = required >= 94.712
                ? 10.278
                : (required - 53.6) / 4;
            return {
                main_l_out: leftOutboard,
                main_l_in: leftInboard,
                main_r_in: leftInboard,
                main_r_out: leftOutboard,
                reserve_1_l: reserve,
                reserve_4_r: reserve,
                center_wing: required
                    - ((leftOutboard + leftInboard) * 2)
                    - (reserve * 2),
            };
        }
        return null;
    };

    const roundBaseDistribution = (aircraftType, planned, required) => {
        if (!planned || !Number.isFinite(required)) return planned;
        const layouts = {
            B757: {pairs: [["left", "right"]], balance: "ctr"},
            B767ER: {pairs: [["left", "right"]], balance: "ctr"},
            A300: {
                pairs: [["l_out", "r_out"], ["l_in", "r_in"]],
                balance: "tt",
            },
            "B747-400": {
                pairs: [
                    ["main_l_out", "main_r_out"],
                    ["main_l_in", "main_r_in"],
                    ["reserve_2_l", "reserve_3_r"],
                ],
                balance: "center_wing",
            },
            "B747-8": {
                pairs: [
                    ["main_l_out", "main_r_out"],
                    ["main_l_in", "main_r_in"],
                    ["reserve_1_l", "reserve_4_r"],
                ],
                balance: "center_wing",
            },
        };
        const layout = layouts[aircraftType];
        if (!layout) return planned;
        const rounded = {...planned};
        layout.pairs.forEach(([left, right]) => {
            const totalTenths = Math.round((planned[left] + planned[right]) * 10);
            rounded[left] = Math.ceil(totalTenths / 2) / 10;
            rounded[right] = Math.floor(totalTenths / 2) / 10;
        });
        rounded[layout.balance] = Math.round((required - Object.entries(rounded)
            .filter(([tank]) => tank !== layout.balance)
            .reduce((sum, [, value]) => sum + value, 0)) * 10) / 10;
        return rounded;
    };

    const planFuelByTank = ({
        aircraftType,
        required,
        remaining,
        actual,
        apuRunning,
        apuAllowance,
        apuSource,
    }) => {
        if (apuRunning === null) return null;
        const planned = roundBaseDistribution(
            aircraftType,
            basePlan(aircraftType, required, remaining, actual),
            required
        );
        if (!planned) return null;
        if (apuRunning) {
            if (!Number.isFinite(apuAllowance) || !(apuSource in planned)) {
                return null;
            }
            planned[apuSource] += apuAllowance;
        }
        return planned;
    };

    const calculateApuAllowance = ({
        plannedDepartureUtc,
        windowMinutes,
        confirmedAtMs,
        rate,
    }) => {
        const departureMs = Date.parse(plannedDepartureUtc);
        if (!Number.isFinite(departureMs) || !Number.isFinite(rate) || rate < 0) {
            return null;
        }
        const effectiveDepartureMs = departureMs + (windowMinutes * 60000);
        const remainingHours = Math.max(
            0,
            (effectiveDepartureMs - confirmedAtMs) / 3600000
        );
        const rawAllowance = remainingHours * rate;
        return Math.ceil((rawAllowance * 20) - 1e-9) / 20;
    };

    const remainingReadingsComplete = (tankCodes, remaining) => (
        tankCodes.length > 0
        && tankCodes.every((tankCode) => Number.isFinite(remaining[tankCode]))
    );

    const api = {
        basePlan,
        calculateApuAllowance,
        planFuelByTank,
        roundBaseDistribution,
        remainingReadingsComplete,
    };
    if (typeof module !== "undefined" && module.exports) {
        module.exports = api;
    }
    if (typeof window !== "undefined") {
        window.NeoScorpionFuelPlanning = api;
    }
    if (typeof document === "undefined") return;

    const sanitizeKlb = (value) => {
        const raw = String(value ?? "");
        if (/[-+eE,\s]/.test(raw)) return "";
        const match = raw.match(/(?:\d+(?:\.\d{0,2})?|\.\d{0,2})/);
        return match ? match[0] : "";
    };
    document.querySelectorAll([
        'input[name="required_fuel"]', 'input[name="inbound_fuel"]',
        'input[name="apu_override_allowance"]',
        'input[name^="remaining_"]', 'input[name^="actual_"]',
        'input[name^="correct_actual_"]',
    ].join(",")).forEach((input) => {
        input.setAttribute("data-k-lb-input", "");
        input.inputMode = "decimal";
        input.addEventListener("input", () => { input.value = sanitizeKlb(input.value); });
    });

    const numberOrNull = (value) => {
        if (String(value ?? "").trim() === "") return null;
        const parsed = Number(value);
        return Number.isFinite(parsed) ? parsed : null;
    };
    const displayFuel = (value) => {
        if (!Number.isFinite(value)) return "INCOMPLETE";
        return `${Math.abs(value * 10 - Math.round(value * 10)) > 1e-9
            ? value.toFixed(2)
            : value.toFixed(1)} K LBS`;
    };
    const displayApuAllowance = (value) => {
        const text = displayFuel(value);
        return text === "INCOMPLETE" ? "APU INCOMPLETE" : `APU ${text.replace(" K LBS", "K")}`;
    };

    document.querySelectorAll("[data-fuel-planning-form]").forEach((form) => {
        const card = form.closest("[data-fuel-assignment-id]");
        const apuRunningInput = form.querySelector("[data-apu-running]");
        const apuSourceInput = form.querySelector("[data-apu-source]");
        const apuSourceWrap = form.querySelector("[data-apu-source-wrap]");
        const apuOverrideEnabled = form.querySelector("[data-apu-override-enabled]");
        const apuOverrideWrap = form.querySelector("[data-apu-override-wrap]");
        const apuOverrideValueWrap = card.querySelector("[data-apu-override-value-wrap]");
        const apuOverrideValue = card.querySelector("[data-apu-override-value]");
        const plannedEntryCue = form.querySelector("[data-planned-entry-cue]");
        if (!card || !apuRunningInput || !apuSourceInput) return;

        const initialApuRunning = form.dataset.initialApuRunning;
        const persistedAllowanceLbs = numberOrNull(
            form.dataset.persistedApuAllowanceLbs
        );
        const persistedAutomaticAllowanceLbs = numberOrNull(
            form.dataset.persistedAutomaticApuAllowanceLbs
        );
        const requiredLbs = numberOrNull(form.dataset.requiredLbs);
        const required = requiredLbs === null ? null : requiredLbs / 1000;
        const rate = Number(form.dataset.effectiveApuRate);
        const windowMinutes = Number(form.dataset.windowMinutes || 0);

        const update = () => {
            const selectedApu = apuRunningInput.value;
            const apuRunning = selectedApu === "yes"
                ? true
                : selectedApu === "no"
                ? false
                : null;
            const sourceRequired = apuRunning === true;
            apuSourceWrap.hidden = !sourceRequired;
            apuSourceInput.disabled = !sourceRequired;
            apuSourceInput.required = sourceRequired;
            if (!sourceRequired) apuSourceInput.value = "";
            if (apuOverrideWrap) apuOverrideWrap.hidden = !sourceRequired;
            const overrideEnabled = apuOverrideEnabled?.value === "1";
            if (apuOverrideValueWrap) {
                apuOverrideValueWrap.hidden = !sourceRequired || !overrideEnabled;
            }
            if (!sourceRequired && apuOverrideEnabled) {
                apuOverrideEnabled.value = "0";
                if (apuOverrideValue) apuOverrideValue.value = "";
            }

            let automaticAllowance = null;
            if (apuRunning === false) {
                automaticAllowance = 0;
            } else if (apuRunning === true) {
                automaticAllowance = (
                    initialApuRunning === "yes" && (
                        persistedAutomaticAllowanceLbs !== null
                        || persistedAllowanceLbs !== null
                    )
                )
                    ? (persistedAutomaticAllowanceLbs ?? persistedAllowanceLbs) / 1000
                    : calculateApuAllowance({
                        plannedDepartureUtc: form.dataset.plannedDepartureUtc,
                        windowMinutes,
                        confirmedAtMs: Date.now(),
                        rate,
                    });
            }
            const overrideAllowance = (
                apuRunning === true && overrideEnabled
                    ? numberOrNull(apuOverrideValue?.value)
                    : null
            );
            const allowance = overrideAllowance ?? automaticAllowance;

            const remaining = {};
            const actual = {};
            const tankCodes = [];
            form.querySelectorAll("[data-fuel-reading]").forEach((input) => {
                const value = numberOrNull(input.value);
                const target = input.dataset.fuelReading === "remaining"
                    ? remaining
                    : actual;
                target[input.dataset.tankCode] = (
                    input.dataset.fuelReading === "actual" && value === null
                        ? 0
                        : value
                );
                if (input.dataset.fuelReading === "remaining") {
                    tankCodes.push(input.dataset.tankCode);
                }
            });
            const remainingComplete = remainingReadingsComplete(tankCodes, remaining);
            const planned = remainingComplete
                ? planFuelByTank({
                    aircraftType: form.dataset.aircraftType,
                    required,
                    remaining,
                    actual,
                    apuRunning,
                    apuAllowance: allowance,
                    apuSource: apuSourceInput.value,
                })
                : null;
            if (plannedEntryCue) plannedEntryCue.hidden = remainingComplete;

            const automaticOutput = card.querySelector("[data-apu-automatic-output]");
            if (automaticOutput) automaticOutput.textContent = displayFuel(automaticAllowance);
            const overrideOutput = card.querySelector("[data-apu-override-output]");
            if (overrideOutput) overrideOutput.textContent = (
                overrideAllowance === null ? "-" : displayFuel(overrideAllowance)
            );
            const allowanceOutput = card.querySelector("[data-apu-allowance-output]");
            if (allowanceOutput) allowanceOutput.textContent = displayApuAllowance(allowance);
            const calculatedReference = card.querySelector("[data-apu-calculated-reference]");
            if (calculatedReference) {
                calculatedReference.textContent = `Calculated: ${displayApuAllowance(automaticAllowance).replace("APU ", "")}`;
            }
            const targetOutput = card.querySelector("[data-fueling-target-output]");
            if (targetOutput) targetOutput.textContent = displayFuel(
                required !== null && allowance !== null ? required + allowance : null
            );
            const selectedSourceOption = apuSourceInput.selectedOptions?.[0];
            const sourceOutput = card.querySelector("[data-apu-source-output]");
            if (sourceOutput) sourceOutput.textContent = (
                sourceRequired && apuSourceInput.value
                    ? selectedSourceOption?.textContent || "-"
                    : "-"
            );
            form.querySelectorAll("[data-planned-tank]").forEach((cell) => {
                const value = planned?.[cell.dataset.plannedTank];
                cell.textContent = Number.isFinite(value)
                    ? value.toFixed(1)
                    : "-";
            });
        };

        form.addEventListener("input", update);
        form.addEventListener("change", update);
        update();
    });
})();

File: app/services/neoscorpion_fuel_planning.py
from decimal import Decimal, ROUND_HALF_UP


def plan_fuel_by_tank(
    aircraft_type,
    required_lbs,
    *,
    remaining_lbs_by_tank=None,
    actual_lbs_by_tank=None,
    apu_running=None,
    apu_allowance_lbs=None,
    apu_source_tank_code=None,
):
    """Return the Hanzo-equivalent planned pounds by tank, or None if incomplete."""
    if required_lbs is None or apu_running is None:
        return None

    required = _decimal(required_lbs)
    remaining = {
        code: _decimal(value or 0)
        for code, value in (remaining_lbs_by_tank or {}).items()
    }
    actual = {
        code: _decimal(value or 0)
        for code, value in (actual_lbs_by_tank or {}).items()
    }
    planners = {
        "B757": _plan_b757,
        "A300": _plan_a300,
        "B767ER": _plan_b767er,
        "B747-400": _plan_b747_400,
        "B747-8": _plan_b747_8,
    }
    planner = planners.get(aircraft_type)
    if planner is None:
        return None

    planned = _round_base_distribution(
        aircraft_type,
        planner(required, remaining, actual),
        required,
    )
    if apu_running:
        if (
            apu_allowance_lbs is None
            or apu_source_tank_code not in planned
        ):
            return None
        planned[apu_source_tank_code] += _decimal(apu_allowance_lbs)
    return planned


def _round_base_distribution(aircraft_type, planned, required):
    """Use operational 0.1K tank increments while preserving Required total."""
    pairs_and_balance = {
        "B757": ((("left", "right"),), "ctr"),
        "B767ER": ((("left", "right"),), "ctr"),
        "A300": (
            (("l_out", "r_out"), ("l_in", "r_in")),
            "tt",
        ),
        "B747-400": (
            (
                ("main_l_out", "main_r_out"),
                ("main_l_in", "main_r_in"),
                ("reserve_2_l", "reserve_3_r"),
            ),
            "center_wing",
        ),
        "B747-8": (
            (
                ("main_l_out", "main_r_out"),
                ("main_l_in", "main_r_in"),
                ("reserve_1_l", "reserve_4_r"),
            ),
            "center_wing",
        ),
    }
    pair_data = pairs_and_balance.get(aircraft_type)
    if pair_data is None:
        return planned
    pairs, balancing_tank = pair_data
    rounded = dict(planned)
    increment = Decimal("100")
    for left_code, right_code in pairs:
        total_increments = int(
            ((planned[left_code] + planned[right_code]) / increment).quantize(
                Decimal("1"),
                rounding=ROUND_HALF_UP,
            )
        )
        left_increments = (total_increments + 1) // 2
        right_increments = total_increments // 2
        rounded[left_code] = increment * left_increments
        rounded[right_code] = increment * right_increments

    rounded[balancing_tank] = _decimal(required) - sum(
        value for tank_code, value in rounded.items() if tank_code != balancing_tank
    )
    return rounded


def _plan_b757(required, remaining, _actual):
    return _plan_three_tank_wing_aircraft(
        required,
        remaining,
        wing_max=Decimal("14600"),
        wing_threshold=Decimal("29200"),
    )


def _plan_b767er(required, remaining, _actual):
    return _plan_three_tank_wing_aircraft(
        required,
        remaining,
        wing_max=Decimal("40200"),
        wing_threshold=Decimal("80400"),
    )


def _plan_three_tank_wing_aircraft(
    required,
    remaining,
    *,
    wing_max,
    wing_threshold,
):
    center_remaining = remaining.get("ctr", Decimal("0"))
    if center_remaining > 0:
        left = (
            wing_max
            if required >= wing_threshold + center_remaining
            else (required - center_remaining) / 2
        )
    else:
        left = wing_max if required >= wing_threshold else required / 2
    center = (
        center_remaining
        if required <= wing_threshold
        else required - left - left
    )
    return {"left": left, "ctr": center, "right": left}


def _plan_a300(required, remaining, _actual):
    center_remaining = remaining.get("ctr", Decimal("0"))
    outboard = (
        Decimal("8200")
        if required >= Decimal("16400")
        else (required - Decimal("4100")) / 2
    )
    if center_remaining > 0 and required < Decimal("78600"):
        inboard = (
            (required - outboard - outboard) / 2
        ) - (center_remaining / 2)
    else:
        inboard = (
            Decimal("31100")
            if required >= Decimal("78600")
            else (required - outboard - outboard) / 2
        )
    center = (
        Decimal("31100")
        if required >= Decimal("109700")
        else required - outboard - inboard - inboard - outboard
    )
    trim = (
        Decimal("0")
        if required <= Decimal("109700")
        else required - outboard - inboard - center - inboard - outboard
    )
    return {
        "l_out": outboard,
        "l_in": inboard,
        "ctr": center,
        "r_in": inboard,
        "r_out": outboard,
        "tt": trim,
    }


def _plan_b747_400(required, _remaining, actual):
    left_outboard = (
        Decimal("29292")
        if required >= Decimal("117168")
        else required / 4
    )
    left_inboard = (
        Decimal("84058")
        if required >= Decimal("244410")
        else actual.get("main_l_in", Decimal("0"))
    )
    main_total = (left_outboard + left_inboard) * 2
    reserve = (
        Decimal("8857")
        if required >= Decimal("163510")
        else (required - main_total) / 2
    )
    center = required - main_total - reserve - reserve
    return {
        "main_l_out": left_outboard,
        "main_l_in": left_inboard,
        "main_r_in": left_inboard,
        "main_r_out": left_outboard,
        "reserve_2_l": reserve,
        "reserve_3_r": reserve,
        "center_wing": center,
    }


def _plan_b747_8(required, _remaining, actual):
    left_outboard = (
        Decimal("35644")
        if required >= Decimal("183688")
        else actual.get("main_l_out", Decimal("0"))
    )
    left_inboard = (
        Decimal("96681")
        if required >= Decimal("285206")
        else actual.get("main_l_in", Decimal("0"))
    )
    reserve = (
        Decimal("10278")
        if required >= Decimal("94712")
        else (required - Decimal("53600")) / 4
    )
    center = required - ((left_outboard + left_inboard) * 2) - (reserve * 2)
    return {
        "main_l_out": left_outboard,
        "main_l_in": left_inboard,
        "main_r_in": left_inboard,
        "main_r_out": left_outboard,
        "reserve_1_l": reserve,
        "reserve_4_r": reserve,
        "center_wing": center,
    }


def _decimal(value):
    return Decimal(str(value))
